src/persist/src/nemesis/direct.rs

```diff
@@ -7,8 +7,93 @@
+//! and read from one (or many) dataflowd processes. The only communication
+//! between these two is via persist.
+//!
+//! [collection]: https://docs.rs/differential-dataflow/0.12.0/differential_dataflow/collection/struct.Collection.html
+//!
+//! The external API guarantees of persist are in terms of its
+//! [StreamReadHandle]s and [StreamWriteHandle]s. Also included is the
+//! [PersistedSource] operator which wraps a StreamReadHandle to give native
+//! differential dataflow semantics.
+//!
+//! The ingestd side uses the read and write handles (technically it currently
+//! uses some extra dataflow operators that wrap them, but ingestd will
+//! eventually get off dataflow so these are going away at some point). We model
+//! this as a set of [RuntimeWorker]s (here [DataflowWorker] is the impl of
+//! that), each concurrently issuing calls to read and write handles and
+//! recording the responses.
+//!
+//! The dataflowd side uses the PersistedSource operator. We model this with a
+//! multi-[timely::worker::Worker] dataflow per persisted collection which
+//! simply captures the output for nemesis validation. These workers are driven
+//! using exactly the [timely::execute] runtime used in production. Additionally
+//! a small bit of [DataflowProgress] tracking is added, which is used by the
+//! ingestd-side to keep traffic more "interesting" by situationally waiting
+//! until the dataflow-side has caught up.
+//!
+//! Not unexpectedly, concurrency was a source of many of persist's initial
+//! bugs, so we make a particular effort to ensure that the production
+//! concurrency is modeled. One tricky thing is that the real ingestd and
+//! dataflowd tie the lifetime of the persist runtime to the lifetime of the
+//! process. However, we want nemesis to model restarts (another likely source
+//! of bugs), which is in tension with maximal concurrency (because persist
+//! requires an exclusive-writer lock, which serializes runtime start/stop
+//! operations). As a compromise, we model the source of truth in [DirectCore]
+//! behind a Mutex, and then each ingestd-side [DataflowWorker] caches the read
+//! and write handles for each collection the first time they are used after a
+//! persist runtime restart. These caches are invalidated using lock-free
+//! atomics to avoid introducing a barrier between the ingestd workers.
+//! [DirectShared] sits in between these two layers and is mostly a helper to
+//! deal with cache invalidation.
+//!
+//! [DirectShared] is also a convenient place to house nemesis's odd split of
+//! getting a snapshot and reading it. We do this split to model a very large
+//! snapshot that takes a long time to consume, so other reads and writes
+//! continue on concurrently in the background.
+//!
+//! A missing piece in all this is that the ingestd and dataflowd processes (and
+//! thus their respective persist runtimes) can be restarted independently.
```

Comment:
is it fair to say that in a platform world, their respective persist runtimes are more like their `PersistClient`s and `persistd` is like the persistence server?

Not trying to be super pedantic, just want to make sure I have the right mental model. My understanding is that everything that's currently part of `RuntimeImpl` will live in `persistd` and and `persistd` will be the only one to write out to storage. Then, instead of the mpsc queue to communicate with `RuntimeImpl` we'll have something like a `ReadPersistClient` or a `ReadWritePersistClient` that connects to `persistd` and those connections could be arbitrarily restarted, which is the thing we want to model, and separately I guess the core persistd could also be restarted.

I realize this wasn't really a question so much as me trying to explain this out to myself, but does that all sound reasonable?

Reply:
Something like that could work, but so far I haven't seen a compelling reason to make persist into a separate service. And there are a couple of compelling reasons to _not_ make it a service: fewer services=operational simplicity and the blue-green/prod-staging story means we need to build a pure read-only tail anyway. The way I (currently) think of this is that persist is a "layer on top of s3" and more specifically a "fat client", which comes in two flavors. So each ingestd process will run a read-write persist runtime and each dataflowd will run a read-only runtime.